RFC: Mostly replace imageParts

mtrmac opened the pull request:
This reduces `imageParts` from various individual fields to a pair of (unnormalized `reference.Named`, `hasRegistry`), with the capability to get a normalized `reference.Named` value (either directly or by supplying the missing registry hostname); and ports most users to use the references.

Notably this removes the `.assemble()` method which could return a `:none` tag for digested inputs, hopefully addressing #2026, and _partially_ #1139 .

This involves _several_ behavior changes; not using the `:none` tag seems clearly desirable, others should be harmless but are less clear:
- The names used during pull output on the console, and returned by `PullFrom…`, now do not include any tag if the user did not specify one. AFAICS that should not break the subsequent image lookup.
- Tagging as `docker.io/busybox:foo` will actually create a `docker.io/library/busybox:foo` tag.
- It’s might now be possible to tag images with a digested reference (previously this created a nonsensical tag which was probably mostly ignored)

Also, some of the old behavior, notably WRT tags, while expunged from `imageParts` proper, has been preserved in a new method `imageParts.suspiciousRefNameTagValuesForSearch`, used in `Image.MatchRepoTag` and `findImageInRepoTags`. The heuristics in both of these methods don’t really make sense to me, but because I don’t understand them, I didn’t touch them. At least the code computing those values is now off the primary code paths.

See the individual commit messages for details.

Absolutely untested, apart from `libpod/image` unit tests.

@baude RFC

rhatdan commented on libpod/image/parts.go:
```diff
@@ -88,6 +89,23 @@ func (ip *imageParts) suspiciousTagValueForSearch() string {
+// qualified with registry.
+func (ip *imageParts) referenceWithRegistry(registry string) (reference.Named, error) {
+	if ip.hasRegistry {
+		return nil, fmt.Errorf("internal error: referenceWithRegistry called on imageParts with a registry (%#v)", *ip)
```
Why doesn't this uses errors.Errorf()

mtrmac replied:
Fixed (in both instances).

rhatdan commented:
Nice work @mtrmac

mheon commented:
`build_each_commit` takes ~11 minutes per 10 commits. We have 30+ commits here, so I suggest bumping the timeout to something like 45 minutes to ensure it has time to finish.

mheon commented:
@mtrmac I agree that the semantics of MatchRepoTag are somewhat questionable, but we couldn't come up with a better way when we originally wrote the code. I think that's probably a separate PR to fix, though.

mtrmac commented:
> `build_each_commit` takes ~11 minutes per 10 commits. We have 30+ commits here, so I suggest bumping the timeout to something like 45 minutes to ensure it has time to finish.

Or, well, I could squash some of the ridiculously small ones :)  If that’s preferred, I can do it tomorrow.

mheon commented:
@mtrmac I don't really mind leaving the commits as-is, though it would speed the tests up a fair bit. If you do choose to hit the timeout, it's in cirrus.yml (the first instance in the file of the string `30m`)

mheon commented:
We're good to review/merge now that 1.0 is branched

baude commented:
can we rebase first?

mtrmac commented:
Rebased, and squashed a little (still leaving separate commits for notable behavior changes). Clearly this could be squashed further.

rhatdan commented:
LGTM
@baude @mheon @vrothberg @giuseppe @TomSweeneyRedHat PTAL

vrothberg commented:
LGTM :+1:

vrothberg commented:
/approve

TomSweeneyRedHat commented:
@mtrmac looks like you need a rebase

mtrmac commented:
Rebased.

@baude #2113 has made `imageParts` public via `DecomposeString`. Given the very dubious value of `tag`, and that the output of `DecomposeString` was not really used anywhere, I have just removed/reverted that part in here.

Feel free to make that public again, but (even if I’m too scared to just drop all existing uses outright)., any net new users of the `tag` value which can go `"none"` should IMHO come with a very strong and documented rationale.

vrothberg commented:
/lgtm
Thanks a lot, @mtrmac !
